File: hw/chavez/u3/hw25TemplateMethodAndObserver/Observer/Java/Observer/src/main/java/ec/edu/espe/observer/model/Stock.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.edu.espe.observer.model;

import java.util.ArrayList;
import java.util.Iterator;

/**
 *
 * @author Odalys Chavez, CodeBreakers, @ESPE
 */
public class Stock {
    protected String symbol;
    protected double price;
    private ArrayList investors = new ArrayList();
    
    public Stock(){
        
    }
    public void addObserver(IInvestor investor) {
        investors.add(investor);
    }
    public void deleteObserver(IInvestor investor) {
        investors.remove(investor);
    }
    public void notifyObservers(Object args) {
        Iterator i = investors.iterator();
        while (i.hasNext()) {
            IInvestor investor = (IInvestor)i.next();
            investor.update(this, args);
        }
    }
    public String getSymbol() {
        return symbol;
    }
}

File: hw/basurto/u3/hw26Observer/hw26Observer.js/model/stock.js
export class Stock {
    constructor() {
        this._symbol = "";
        this._price = 0.0;
        this._investors = [];
    }

    addObserver(investor) {
        this._investors.push(investor);
    }

    deleteObserver(investor) {
        this._investors = this._investors.filter(inv => inv !== investor);
    }

    notifyObservers(args) {
        this._investors.forEach(investor => {
            investor.update(this, args);
        });
    }

    get symbol() {
        return this._symbol;
    }

    set symbol(value) {
        this._symbol = value;
        this.notifyObservers(value);
    }

    get price() {
        return this._price;
    }

    set price(value) {
        this._price = value;
        this.notifyObservers(value);
    }
}